### PR TITLE
Fix invalid tokens

### PR DESCRIPTION
**Pull request by IvanBorislavovDimitrov**
If we upload large archives sometimes the deployment fails because the session expires and csrf tokens get invalid.

### CODE REVIEW

File: clients/baseclient/client_util.go [hunk 1]
```diff
@@ -23,7 +23,11 @@ func shouldRetry(err error) bool {
 	if err == nil {
 		return false
 	}
-	isMatching, _ := regexp.MatchString(" EOF$", err.Error())
+	isMatching := strings.Contains(err.Error(), "retry is needed")
```
Comment (enchobelezirev) [inline clients/baseclient/client_util.go]: Maybe the 2 checks could be extracted in a separate methods for better readability?

File: clients/csrf/csrf_token_fetcher.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package csrf
+
+import "net/http"
+
+type CsrfTokenFetcher interface {
+	FetchNewCsrfToken(url string, currentRequest *http.Request) (string, string, error)
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_fetcher.go]: Personally, I do not like more than 3 args if they needed indeed. Could a return object be introduced here and return it directly?

File: clients/csrf/csrf_token_fetcher_impl.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+func NewCsrfTokenFetcherImpl(transport *Transport) *CsrfTokenFetcherImpl {
+	return &CsrfTokenFetcherImpl{transport: transport}
+}
+
+func (c *CsrfTokenFetcherImpl) FetchNewCsrfToken(url string, currentRequest *http.Request) (string, string, error) {
+	fetchTokenRequest, _ := http.NewRequest(http.MethodGet, url, nil)
+	fetchTokenRequest.Header.Set(XCsrfToken, CsrfTokenHeaderFetchValue)
+	fetchTokenRequest.Header.Set("Content-Type", "application/json")
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_fetcher_impl.go]: "application/json" -> constant

File: clients/csrf/csrf_token_fetcher_impl.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+type CsrfTokenFetcherImpl struct {
+	transport *Transport
+}
+
+func NewCsrfTokenFetcherImpl(transport *Transport) *CsrfTokenFetcherImpl {
+	return &CsrfTokenFetcherImpl{transport: transport}
+}
+
+func (c *CsrfTokenFetcherImpl) FetchNewCsrfToken(url string, currentRequest *http.Request) (string, string, error) {
+	fetchTokenRequest, _ := http.NewRequest(http.MethodGet, url, nil)
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_fetcher_impl.go]: Why ignoring the error?

File: clients/csrf/csrf_token_fetcher_impl.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+type CsrfTokenFetcherImpl struct {
+	transport *Transport
+}
+
+func NewCsrfTokenFetcherImpl(transport *Transport) *CsrfTokenFetcherImpl {
+	return &CsrfTokenFetcherImpl{transport: transport}
+}
+
+func (c *CsrfTokenFetcherImpl) FetchNewCsrfToken(url string, currentRequest *http.Request) (string, string, error) {
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_fetcher_impl.go]: Rename FetchNewCsrfToken -> FetchCsrfToken

File: clients/csrf/csrf_token_fetcher_impl.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	fetchTokenRequest, _ := http.NewRequest(http.MethodGet, url, nil)
+	fetchTokenRequest.Header.Set(XCsrfToken, CsrfTokenHeaderFetchValue)
+	fetchTokenRequest.Header.Set("Content-Type", "application/json")
+
+	cliConnection := plugin.NewCliConnection(os.Args[1])
+	token, _ := cliConnection.AccessToken()
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_fetcher_impl.go]: Again ignoring the error?

File: clients/csrf/csrf_token_fetcher_impl.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	fetchTokenRequest.Header.Set("Authorization", token)
+	NewCookiesUpdater(currentRequest.Cookies(), fetchTokenRequest).updateCookiesIfNeeded()
+
+	response, err := c.transport.Transport.RoundTrip(fetchTokenRequest)
+	if err != nil {
+		return "", "", err
+	}
+	if len(response.Cookies()) != 0 {
+		fetchTokenRequest.Header.Del(CookieHeader)
+		NewCookiesUpdater(response.Cookies(), fetchTokenRequest).updateCookiesIfNeeded()
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_fetcher_impl.go]: Wouldn't it be better if the fetched has a CookieUpdater inside itself?

File: clients/csrf/csrf_token_fetcher_impl.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+		c.transport.Csrf.Cookies = fetchTokenRequest.Cookies()
+
+		response, err = c.transport.Transport.RoundTrip(fetchTokenRequest)
+	}
+
+	if err != nil {
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_fetcher_impl.go]: Why this check is here?

File: clients/csrf/csrf_token_fetcher_impl.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+		return "", "", err
+	}
+
+	return response.Header.Get(XCsrfHeader), response.Header.Get(XCsrfToken), nil
+}
+
+func getFetchNewTokenUrl(req *http.Request) string {
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_fetcher_impl.go]: Is this a good naming for a method "getFetch..."?

File: clients/csrf/csrf_token_manager.go [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package csrf
+
+import "net/http"
+
+type CsrfTokenManager interface {
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_manager.go]: *Manager as a suffix is not better for naming of classes or interfaces. This is the same as God. Try to rename it if the interface is referring to something more and you need to add "and" in the name, then separate the interface into two :)

File: clients/csrf/transport.go [hunk 11]
```diff
@@ -19,16 +21,27 @@ type Transport struct {
...
 	}
-	csrfHeader, csrfToken := res.Header.Get("X-Csrf-Header"), res.Header.Get("X-Csrf-Token")
-	if csrfHeader != "" && csrfToken != "" {
-		t.Csrf.Header, t.Csrf.Token = csrfHeader, csrfToken
+	isRetryNeeded, err := csrfTokenManager.isRetryNeeded(res)
+	if err != nil {
+		return nil, err
 	}
+
+	if isRetryNeeded {
```
Comment (enchobelezirev) [inline clients/csrf/transport.go]: I do not get this? Why on retryNeed you will return the response?

File: clients/csrf/csrf_token_manager.go [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+	setCsrfToken() error
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_manager.go]: Such naming conventions are great but they need to be follow by some method arguments to support them: setCsrfToken(<where-to-set-it>)

File: clients/csrf/csrf_token_manager.go [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+	initializeToken(force bool, url string) error
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_manager.go]: force method param does not mean anything for a person looking at the code for the first time.

File: clients/csrf/csrf_token_manager_impl.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+	if err != nil {
+		return err
+	}
+
+	c.updateCurrentCsrfTokens(c.request, c.transport)
+
+	return nil
+}
+
+func (c *CsrfTokenManagerImpl) initializeToken(force bool, url string) error {
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_manager_impl.go]: Why is the force needed?

File: clients/mtaclient/retryable_mta_rest_client.go [hunk 15]
```diff
@@ -117,7 +117,9 @@ func (c RetryableMtaRestClient) StartMtaOperation(operation models.Operation) (R
 }
 func (c RetryableMtaRestClient) UploadMtaFile(file os.File) (*models.FileMetadata, error) {
 	uploadMtaFileCb := func() (interface{}, error) {
-		return c.mtaClient.UploadMtaFile(file)
+		reopenedFile, _ := os.Open(file.Name())
```
Comment (enchobelezirev) [inline clients/mtaclient/retryable_mta_rest_client.go]: error ignore again ...
Who is responsible for closing the file here?

File: clients/csrf/forbidden_error.go [hunk 16]
```diff
@@ -0,0 +1,12 @@
+package csrf
+
+const RetryIsNeeded = "retry is needed"
+
+type ForbiddenError struct {
```
Comment (enchobelezirev) [inline clients/csrf/forbidden_error.go]: Do not understand at all.

File: clients/csrf/fakes/fake_csrf_token_manager.go [hunk 17]
```diff
@@ -0,0 +1,17 @@
+package fakes
+
+import "net/http"
+
+const FakeCsrfTokenHeader = "fake-xcsrf-token-header"
+const FakeCsrfTokenValue = "fake-xcsrf-token-value"
+
+type FakeCsrfTokenFetcher struct {
```
Comment (enchobelezirev) [inline clients/csrf/fakes/fake_csrf_token_manager.go]: How you generated this struct?

File: clients/csrf/csrf_token_manager_impl.go [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+			}
+
+			return c.transport.Csrf.Token != "", nil
+		}
+	}
+
+	return false, nil
+}
+
+func (c *CsrfTokenManagerImpl) updateCurrentCsrfTokens(request *http.Request, t *Transport) {
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_manager_impl.go]: updateCurrentCsrfTokens isn't it just one token?

File: clients/csrf/csrf_token_manager_impl.go [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func (c *CsrfTokenManagerImpl) initializeToken(force bool, url string) error {
+	if force || !c.transport.Csrf.IsInitialized {
+		var err error
+		c.transport.Csrf.Header, c.transport.Csrf.Token, err = c.csrfTokenFetcher.FetchNewCsrfToken(url, c.request)
+		if err != nil {
+			return err
+		}
+		c.transport.Csrf.IsInitialized = true
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_manager_impl.go]: I am not happy with such "global" state variables

File: clients/baseclient/client_util.go [hunk 20]
```diff
@@ -23,7 +23,11 @@ func shouldRetry(err error) bool {
...
+	if isMatching {
+		return true
+	}
+	isMatching, _ = regexp.MatchString(" EOF$", err.Error())
```
Comment (enchobelezirev) [inline clients/baseclient/client_util.go]: Error ignore?

File: clients/csrf/transport.go [hunk 21]
```diff
@@ -1,14 +1,16 @@
...
 	"github.com/jinzhu/copier"
+	"net/http"
 )
 
 type Csrf struct {
-	Header string
-	Token  string
+	Header              string
+	Token               string
+	IsInitialized       bool
```
Comment (enchobelezirev) [inline clients/csrf/transport.go]: Is this needed?

File: clients/csrf/transport.go [hunk 22]
```diff
@@ -1,34 +1,158 @@
...
+	"os"
 )
 
 type Csrf struct {
-	Header string
-	Token  string
+	Header        string
+	Token         string
+	IsInitialized bool
+	cookies       []*http.Cookie
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply clients/csrf/transport.go]: Moved in a new struct, in the transport.

File: clients/baseclient/client_util.go (hunk 1 above)
Comment (IvanBorislavovDimitrov) [reply clients/baseclient/client_util.go]: Refactored -> new method

File: clients/csrf/csrf_token_fetcher.go (hunk 2 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_fetcher.go]: Fixed, a new object is returned

File: clients/csrf/csrf_token_fetcher_impl.go (hunk 3 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_fetcher_impl.go]: Done

File: clients/csrf/csrf_token_fetcher_impl.go (hunk 4 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_fetcher_impl.go]: Not ignored anymore.

File: clients/csrf/csrf_token_fetcher_impl.go (hunk 5 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_fetcher_impl.go]: Done

File: clients/csrf/csrf_token_fetcher_impl.go (hunk 6 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_fetcher_impl.go]: Not ignored anymore.

File: clients/csrf/csrf_token_fetcher_impl.go (hunk 7 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_fetcher_impl.go]: Done

File: clients/csrf/csrf_token_fetcher_impl.go (hunk 8 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_fetcher_impl.go]: Moved

File: clients/csrf/csrf_token_fetcher_impl.go (hunk 9 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_fetcher_impl.go]: Renamed to -> getCsrfTokenUrl

File: clients/csrf/csrf_token_manager.go (hunk 10 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_manager.go]: Renamed to -> Updater, that's the main functionality of this object.

File: clients/csrf/transport.go (hunk 11 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/transport.go]: It isn't needed, removed

File: clients/csrf/csrf_token_manager.go (hunk 12 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_manager.go]: Renamed to -> checkAndUpdateCsrfToken

File: clients/csrf/csrf_token_manager.go (hunk 13 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_manager.go]: Renamed to -> forceInitializing

File: clients/csrf/csrf_token_manager_impl.go (hunk 14 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_manager_impl.go]: It's needed when we've initialized a csrf token and a new one is needed.

File: clients/mtaclient/retryable_mta_rest_client.go (hunk 15 above)
Comment (IvanBorislavovDimitrov) [reply clients/mtaclient/retryable_mta_rest_client.go]: Not ignored anymore. The file is closed in the file uploader -> uploadPartFile.

File: clients/csrf/forbidden_error.go (hunk 16 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/forbidden_error.go]: Custom error. Thrown when retry is needed.

File: clients/csrf/fakes/fake_csrf_token_manager.go (hunk 17 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/fakes/fake_csrf_token_manager.go]: What do you mean?

File: clients/csrf/csrf_token_manager_impl.go (hunk 18 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_manager_impl.go]: Yes, it is, wrong name...

File: clients/csrf/csrf_token_manager_impl.go (hunk 19 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_manager_impl.go]: There is no other way to preserve the initial state.

File: clients/baseclient/client_util.go (hunk 20 above)
Comment (IvanBorislavovDimitrov) [reply clients/baseclient/client_util.go]: It was ignored so far. This error can occur only if the regex is invalid.

File: clients/csrf/transport.go (hunk 21 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/transport.go]: Yes, you need to know if the csrf is initialized or not.

File: clients/baseclient/client_util.go [hunk 23]
```diff
@@ -38,6 +37,16 @@ func shouldRetry(err error) bool {
 	return false
 }
 
+func isMatching(err error) bool {
+	isMatching := strings.Contains(err.Error(), "retry is needed")
+	if isMatching {
+		return true
+	}
+	isMatching, _ = regexp.MatchString(" EOF$", err.Error())
```
Comment (enchobelezirev) [inline clients/baseclient/client_util.go]: error ignoration?

File: clients/csrf/csrf_paramters/csrf_parameters.go [hunk 24]
```diff
@@ -0,0 +1,6 @@
+package csrf_paramters
+
+type CsrfParameters struct {
```
Comment (enchobelezirev) [inline clients/csrf/csrf_paramters/csrf_parameters.go]: The whole CSRF thing is a request header. Its values are header name and header value

File: clients/csrf/csrf_token_fetcher_impl.go [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+)
+
+const CsrfTokenHeaderFetchValue = "Fetch"
+const CsrfTokensApi = "/api/v1/csrf-token"
+const ContentTypeHeader = "Content-Type"
+const AuthorizationHeader = "Authorization"
+const ApplicationJsonContentType = "application/json"
+const CookieHeader = "CookieHeader"
+
+type CsrfTokenFetcherImpl struct {
```
Comment (enchobelezirev) [inline clients/csrf/csrf_token_fetcher_impl.go]: *Impl is not a good naming, IMO. Maybe DefaultCsrfTokenFetcher?

File: clients/baseclient/client_util.go (hunk 1 above)
Comment (enchobelezirev) [reply clients/baseclient/client_util.go]: Do not see this new method.

File: clients/baseclient/client_util.go (hunk 23 above)
Comment (IvanBorislavovDimitrov) [reply clients/baseclient/client_util.go]: This error can occur only if the regex is invalid.

File: clients/csrf/csrf_paramters/csrf_parameters.go (hunk 24 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_paramters/csrf_parameters.go]: Renamed to -> CsrfRequestHeader

File: clients/csrf/csrf_token_fetcher_impl.go (hunk 25 above)
Comment (IvanBorislavovDimitrov) [reply clients/csrf/csrf_token_fetcher_impl.go]: Renamed to -> DefaultCsrfTokenFetcher

File: commands/mta_operations_command.go [hunk 26]
```diff
@@ -24,7 +24,7 @@ func (c *MtaOperationsCommand) GetPluginCommand() plugin.Command {
 		UsageDetails: plugin.Usage{
 			Usage: "cf mta-ops [-u URL] [--last NUM] [--all]",
 			Options: map[string]string{
-				"u": "Deploy service URL, by default 'deploy-service.<system-domain>'",
+				"u":                         "Deploy service URL, by default 'deploy-service.<system-domain>'",
```
Comment (theghost5800) [inline commands/mta_operations_command.go]: Is it really needed to add spaces?

File: commands/undeploy_command.go [hunk 27]
```diff
@@ -38,11 +38,11 @@ func (c *UndeployCommand) GetPluginCommand() plugin.Command {
    Perform action on an active undeploy operation
    cf undeploy -i OPERATION_ID -a ACTION [-u URL]`,
 			Options: map[string]string{
-				deployServiceURLOpt: "Deploy service URL, by default 'deploy-service.<system-domain>'",
-				operationIDOpt:      "Active undeploy operation id",
-				actionOpt:           "Action to perform on the active undeploy operation (abort, retry, monitor)",
-				forceOpt:            "Force undeploy without confirmation",
-				util.GetShortOption(deleteServicesOpt):             "Delete services",
+				deployServiceURLOpt:                    "Deploy service URL, by default 'deploy-service.<system-domain>'",
```
Comment (theghost5800) [inline commands/undeploy_command.go]: spaces?

File: clients/csrf/default_csrf_token_fetcher.go [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+func NewDefaultCsrfTokenFetcher(transport *Transport) *DefaultCsrfTokenFetcher {
+	return &DefaultCsrfTokenFetcher{transport: transport}
+}
+
+func (c *DefaultCsrfTokenFetcher) FetchCsrfToken(url string, currentRequest *http.Request) (*csrf_paramters.CsrfRequestHeader, error) {
+	fetchTokenRequest, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+	fetchTokenRequest.Header.Set(XCsrfToken, CsrfTokenHeaderFetchValue)
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply clients/csrf/default_csrf_token_fetcher.go]: I think it is required by the backend.

File: commands/mta_operations_command.go (hunk 26 above)
Comment (IvanBorislavovDimitrov) [reply commands/mta_operations_command.go]: They are added from the IDE-> removed

File: commands/undeploy_command.go (hunk 27 above)
Comment (IvanBorislavovDimitrov) [reply commands/undeploy_command.go]: Removed

**Pull request merged**